#166790089 implement social login

shaluchandwani opened the pull request:
#### What does this PR do?
- Added scripts in package
- Added API, Routes, Middleware and helpers for the social login
- Added Tests for the feature 

#### Description of Task to be completed?
Enable Social Login via Google, Facebook, and Twitter.

#### How should this be manually tested?
This can be tested by running "npm run test" and on the browser using the following URL - 
1. http://127.0.0.1:3000/api/social/login/facebook
2.http://127.0.0.1:3000/api/social/login/google
3.http://127.0.0.1:3000/api/social/login/twitter

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/166790089

#### Screenshots (if appropriate)
<img width="1393" alt="Screen Shot 2019-07-03 at 15 23 22" src="https://user-images.githubusercontent.com/51748364/60595162-bba81e80-9da6-11e9-9492-967b77ecc69f.png">
<img width="909" alt="Screen Shot 2019-07-03 at 15 24 18" src="https://user-images.githubusercontent.com/51748364/60595170-c2369600-9da6-11e9-88e7-f9604e831b1c.png">

#### Questions:
N/A

shaluchandwani committed:
ft(social_login): implement social login

- added scripts in package
- Added API , Routes , Middleware and helpers for the social login

[(Finishes #166790089]

kipropbrian commented on src/test/authTest.js:
```diff
@@ -0,0 +1,102 @@
+  //       expect(res.body).to.be.an('object');
+  //       done();
+  //     });
+  // });
```
So many comments. Kindly remove them if you are actually not using them

shaluchandwani replied:
They are tests for Twitter, I will use them as soon as I will get the secret key from them. That is the reason I commented it.

FlevianK replied:
Have them somewhere in your machine, they you can add them later when needed.

FlevianK commented on package.json:
```diff
@@ -30,25 +32,28 @@
     "sequelize-cli": "^5.5.0",
-    "swagger-ui-express": "^4.0.6"
+    "swagger-ui-express": "^4.0.6",
+    "chai": "^4.2.0",
```
This should be under devDependencies

FlevianK commented on package.json:
```diff
@@ -30,25 +32,28 @@
-    "swagger-ui-express": "^4.0.6"
+    "swagger-ui-express": "^4.0.6",
+    "chai": "^4.2.0",
+    "chai-http": "^4.3.0"
```
This should be under devDependencies

FlevianK commented on src/controllers/socialLogin.js:
```diff
@@ -0,0 +1,74 @@
+      mail: req.user.emails[0].value,
+      image: req.user.photos[0].value,
+    });
+    if (newUser) {
```
can we extract this into a function since it is repetitive

shaluchandwani replied:
It is not repetitive, therefore I used it differently. The response is different for facebook and google.

FlevianK commented on src/middleware/socialAcctExists.js:
```diff
@@ -0,0 +1,55 @@
+      } = currentUser[0].dataValues;
+      return res.status(200).json({
+        message: `Welcome to Authors Haven ${displayName} `,
+        data: {
```
Why is this object not the same as the object in line 22?

FlevianK replied:
If they are the same, extract if statement into a function to avoid repeating code.

shaluchandwani replied:
No they are not the same response from Google and Facebook was the same that is why I used it in a single function but Twitter has a different response.

FlevianK commented on src/middleware/socialAcctExists.js:
```diff
@@ -0,0 +1,55 @@
+  },
+  async twitter(req, res, next) {
+    const { displayName } = req.user;
+    const currentUser = await models.user.findAll({
```
Here you are just checking if that user exists, use fineOne in stead of findAll

FlevianK commented on src/middleware/socialAcctExists.js:
```diff
@@ -0,0 +1,55 @@
+const userExists = {
+  async google(req, res, next) {
+    const { emails, displayName } = req.user;
+    const currentUser = await models.user.findAll({
```
Here you are just checking if that user exists, use fineOne in stead of findAll

shaluchandwani replied:
findOne was not working as I have the same email to get login again. I tried it was failing as i am checking length also.

FlevianK commented on src/helpers/tokenGenerator.js:
```diff
@@ -0,0 +1,20 @@
+
+dotenv.config();
+
+class Tokenizer {
```
class name is supposed to be a noun

FlevianK commented on src/controllers/socialLogin.js:
```diff
@@ -0,0 +1,74 @@
+import tokenGeneration from '../helpers/tokenGenerator';
+
+const userInfo = {
+  async googleLogin(req, res) {
```
Function name should start with a verb. Change on all of them

FlevianK commented on src/config/passport.js:
```diff
@@ -0,0 +1,46 @@
+              callbackURL: `${process.env.BASE_URL}/api/social/login/twitter/redirect`
+            },
+            (token, tokenSecret, profile, done) => {
+              console.log(token);
```
why are you pushing console.log?

FlevianK commented on .envsample:
```diff
@@ -1,3 +1,10 @@
 DEV_DATABASE_URL=postgres://john:doe@127.0.0.1:5432/userschema
 TEST_DATABASE_URL=postgres://john:doe@127.0.0.1:5432/userschema_test
 DATABASE_URL=postgres://john:doe@127.0.0.1:5432/userschema
+FCBK_ID=
```
add sample value

FlevianK commented on package.json:
```diff
@@ -7,6 +7,7 @@
     "start": "npm run create && NODE_ENV=production babel-node ./src/index.js",
     "dev": "nodemon --exec babel-node ./src/index.js",
     "sequelize": "sequelize db:migrate",
+    "sequelizeall": "sequelize db:seed:all",
```
remove this script and use migrate. Replace `sequelize` with `migrate` in the travis file

FlevianK commented on package.json:
```diff
@@ -15,7 +16,8 @@
     "lint": "eslint ./",
-    "lint-fix": "eslint ./ --fix"
+    "lint-fix": "eslint ./ --fix",
+    "destroy": "babel-node src/helpers/destroyUsers"
```
remove this too

FlevianK commented on .travis.yml:
```diff
@@ -7,6 +7,8 @@ env:
 before-script:
   - npm add sequelize-cli
   - npm run sequelize
+  - npm destroy
```
remove this

FlevianK commented on .travis.yml:
```diff
@@ -7,6 +7,8 @@ env:
   - npm add sequelize-cli
   - npm run sequelize
+  - npm destroy
+  - npm run sequelizeall
```
remove this

FlevianK commented on .travis.yml:
```diff
@@ -7,6 +7,8 @@ env:
 before-script:
   - npm add sequelize-cli
   - npm run sequelize
```
replace this with `migrate`

FlevianK commented on src/helpers/destroyUsers.js:
```diff
@@ -0,0 +1,6 @@
+import models from '../models';
```
this file is note needed

FlevianK commented on src/middleware/socialAcctExists.js:
```diff
@@ -0,0 +1,55 @@
+import models from '../models';
+import tokenGeneration from '../helpers/tokenGenerator';
+
+// eslint-disable-next-line no-unused-vars
```
remove this, ensure you don't have unused  vars/const

FlevianK commented on src/migrations/20190624-create-userschema.js:
```diff
@@ -1,50 +1,46 @@
+      type: Sequelize.DATE
+    }
+  }),
+  // eslint-disable-next-line no-unused-vars
```
why are you disabling eslint?

FlevianK commented on src/routes/api/socialRout.js:
```diff
@@ -0,0 +1,28 @@
+const socialRout = Router();
+const { google, twitter } = socialAccount;
+
+// social login testing routes
```
Remove this, i.e line 12 to 14. You are supposed to test the actual routes

FlevianK commented on src/routes/api/socialRout.js:
```diff
@@ -0,0 +1,28 @@
+import socialAccount from '../../middleware/socialAcctExists';
+import socialMiddleware from '../../middleware/socialTests';
+
+const socialRout = Router();
```
It should be socialRoute, add an `e` at the end

FlevianK commented on src/test/authTest.js:
```diff
@@ -0,0 +1,102 @@
+      },
+      truncate: false
+    });
+  });
```
add new line

FlevianK commented on src/test/authTest.js:
```diff
@@ -0,0 +1,102 @@
+        expect(res.body).to.be.an('object');
+        done();
+      });
+  });
```
Add new line after every test function

FlevianK commented on src/test/authTest.js:
```diff
@@ -0,0 +1,102 @@
+        done();
+      });
+  });
+  // it('should let a user log in with twitter, test! ', (done) => {
```
Remove commented out code

FlevianK commented on src/test/authTest.js:
```diff
@@ -0,0 +1,102 @@
+        done();
+      });
+  });
+  it('should let save a user if he is already in the database, test! ', (done) => {
```
I don't understand this, review all your descriptions.

FlevianK commented on src/test/authTest.js:
```diff
@@ -0,0 +1,102 @@
+        email: 'luvableshalu@gmail.com',
+      })
+      .end((err, res) => {
+        expect(res.status).to.equal(201);
```
Check out HTTP status code and use the appropriate one, 201 is for `create` action.

FlevianK commented on src/middleware/socialTests.js:
```diff
@@ -0,0 +1,15 @@
+  req.user = {
+    id: req.body.id,
+    displayName: 'shalu chandwani',
+    name: { familyName: 'chandwani', givenName: 'shalu' },
```
why are you using objects or arrays?

FlevianK reviewed (commented):
Other than those reviews, LGTM.

shaluchandwani committed:
ft(social_login): implement social login

- Added the suggested changes
- Added Tests
- Added Swagger Documentation

[(Finishes #166790089]

FlevianK reviewed (changes requested):
Kindly write test for the lines outlined in the test coverage. Those files were introduced by you and must be tested.

Looking forward for 90% and above coverage

shaluchandwani commented:
Please check 
<img width="878" alt="Screen Shot 2019-07-09 at 09 30 12" src="https://user-images.githubusercontent.com/51748364/60868111-95fa9980-a22c-11e9-8b77-c5f7006fe25a.png">

shaluchandwani commented:
Please check i have resolved the change requests
<img width="630" alt="Screen Shot 2019-07-09 at 13 34 32" src="https://user-images.githubusercontent.com/51748364/60884859-5a70c700-a24e-11e9-8959-d19d1929bb6f.png">
